rusoto/core/src/request.rs

```diff
@@ -156,42 +156,46 @@ impl HttpResponse {
 
 #[derive(Clone, Debug, PartialEq)]
 /// An error produced when sending the request, such as a timeout error.
-pub struct HttpDispatchError {
-    message: String,
-}
-
-impl HttpDispatchError {
-    /// Construct a new HttpDispatchError for testing purposes
-    pub fn new(message: String) -> HttpDispatchError {
-        HttpDispatchError { message }
-    }
+pub enum HttpDispatchError {
+    /// Request timed out
+    Timeout,
+    /// Error from client future
+    ClientFutureError (String),
+    /// Internal error in deadline implementation.
+    DeadlineError (String),
+    /// Error from hyper
+    HyperError (String),
+    /// Io error
+    IoError (String),
 }
 
 impl Error for HttpDispatchError {
     fn description(&self) -> &str {
-        &self.message
+        match self {
+            Self::Timeout => "Request timed out",
+            Self::ClientFutureError(s)
+            | Self::DeadlineError(s)
+            | Self::HyperError (s)
+            | Self::IoError (s) => &s,
+        }
     }
 }
 
 impl fmt::Display for HttpDispatchError {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.message)
+        write!(f, "{}", self.description())
     }
 }
 
 impl From<HyperError> for HttpDispatchError {
     fn from(err: HyperError) -> HttpDispatchError {
-        HttpDispatchError {
-            message: err.to_string(),
-        }
+        HttpDispatchError::HyperError (err.to_string())
```

Comment:
Agh, I didn't click buttons in the right order on the review interface.

Should we pass errors like `hyper::Error` through so that those can be inspected as well?

If we're converting errors into string descriptions solely to support `Error::description`, should we just drop that method implementation, since [it's soft-deprecated in favor of Display](https://doc.rust-lang.org/1.38.0/std/error/trait.Error.html#method.description)?

Reply:
Yes I agree. But I have this problem
```
error[E0277]: the trait bound `hyper::Error: std::clone::Clone` is not satisfied
   --> rusoto/core/src/request.rs:167:17
    |
167 |     HyperError (hyper::Error),
    |                 ^^^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `hyper::Error`
    |
    = note: required by `std::clone::Clone::clone`

error[E0369]: binary operation `==` cannot be applied to type `hyper::Error`
   --> rusoto/core/src/request.rs:167:17
    |
167 |     HyperError (hyper::Error),
    |                 ^^^^^^^^^^^^
    |
    = note: an implementation of `std::cmp::PartialEq` might be missing for `hyper::Error`

error[E0369]: binary operation `!=` cannot be applied to type `hyper::Error`
   --> rusoto/core/src/request.rs:167:17
    |
167 |     HyperError (hyper::Error),
    |                 ^^^^^^^^^^^^
    |
    = note: an implementation of `std::cmp::PartialEq` might be missing for `hyper::Error`
```
and same with at least `tokio::timer::timeout::Error`. Should we remove `Clone` and `PartialEq` from `HttpDispatchError`?

Reply:
I've been going back and forth on this since I got this response -- I think because of this we should keep it as-is for now, and I'm going to open an issue to figure out the best path forward here.

Reply:
Alright. I wonder whether we could get around it by using wrapper like `Arc<_>` or `Arc<Box<_>>`

Reply:
`Arc<_>` is probably right, I want to get a second opinion from @matthewkmayer though. :)

Reply:
it only solves `Clone` though

Reply:
`PartialEq` does not seem necessary for tests for me, except perhaps for test cases.